File: virtual interview/task 1/task 1/Heap.java
import java.util.ArrayList;

public class Heap<T extends Comparable<T>> {

    // Each parent has 2^childrenExponent children
    private final int childCount;
    private final ArrayList<T> data;

    public Heap(int childrenExponent) {
        if (childrenExponent < 0) {
            throw new IllegalArgumentException(
                    "childrenExponent must be non-negative");
        }

        this.childCount = 1 << childrenExponent; // 2^x
        this.data = new ArrayList<>();
    }

    // Insert an item into the heap
    public void insert(T item) {
        data.add(item);

        int currentIndex = data.size() - 1;

        while (currentIndex > 0) {
            int nextIndex = swapUp(currentIndex);

            if (nextIndex == currentIndex) {
                break;
            }

            currentIndex = nextIndex;
        }
    }

    // Move item upward if needed
    private int swapUp(int childIndex) {
        int parentIndex = (childIndex - 1) / childCount;

        T childValue = data.get(childIndex);
        T parentValue = data.get(parentIndex);

        if (childValue.compareTo(parentValue) > 0) {
            data.set(parentIndex, childValue);
            data.set(childIndex, parentValue);
            return parentIndex;
        }

        return childIndex;
    }

    // Remove and return maximum element
    public T popMax() {
        if (data.isEmpty()) {
            return null;
        }

        T maxItem = data.get(0);

        // Only one element
        if (data.size() == 1) {
            data.remove(0);
            return maxItem;
        }

        T lastItem = data.remove(data.size() - 1);
        data.set(0, lastItem);

        int currentIndex = 0;

        while (true) {
            int nextIndex = swapDown(currentIndex);

            if (nextIndex == currentIndex) {
                break;
            }

            currentIndex = nextIndex;
        }

        return maxItem;
    }

    // Move item downward if needed
    private int swapDown(int parentIndex) {
        int largestChildIndex = -1;
        T largestChildValue = null;

        int firstChildIndex = parentIndex * childCount + 1;

        for (int i = 0; i < childCount; i++) {
            int childIndex = firstChildIndex + i;

            if (childIndex >= data.size()) {
                break;
            }

            T childValue = data.get(childIndex);

            if (largestChildValue == null ||
                    childValue.compareTo(largestChildValue) > 0) {
                largestChildValue = childValue;
                largestChildIndex = childIndex;
            }
        }

        if (largestChildIndex == -1) {
            return parentIndex;
        }

        T parentValue = data.get(parentIndex);

        if (parentValue.compareTo(largestChildValue) < 0) {
            data.set(parentIndex, largestChildValue);
            data.set(largestChildIndex, parentValue);
            return largestChildIndex;
        }

        return parentIndex;
    }

    // Peek maximum element
    public T peekMax() {
        if (data.isEmpty()) {
            return null;
        }

        return data.get(0);
    }

    // Heap size
    public int size() {
        return data.size();
    }

    // Check if empty
    public boolean isEmpty() {
        return data.isEmpty();
    }

    // Display heap contents
    public void printHeap() {
        System.out.println(data);
    }

    // Test Program
    public static void main(String[] args) {

        // childrenExponent = 2
        // Each parent has 2^2 = 4 children
        Heap<Integer> heap = new Heap<>(2);

        heap.insert(10);
        heap.insert(40);
        heap.insert(15);
        heap.insert(70);
        heap.insert(20);
        heap.insert(100);
        heap.insert(5);
        heap.insert(90);

        System.out.println("Heap:");
        heap.printHeap();

        System.out.println("\nElements removed in descending order:");

        while (!heap.isEmpty()) {
            System.out.print(heap.popMax() + " ");
        }
    }
}

File: July/virtual interview/task 1/task 1/heap.js
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public class PowerOfTwoMaxHeap {

    // Each node has 2^childrenExponent children
    private final int childrenExponent;
    private final int branchingFactor;

    private final List<Integer> heap;

    public PowerOfTwoMaxHeap(int childrenExponent) {
        if (childrenExponent < 0) {
            throw new IllegalArgumentException(
                "childrenExponent must be non-negative"
            );
        }

        this.childrenExponent = childrenExponent;
        this.branchingFactor = 1 << childrenExponent;
        this.heap = new ArrayList<>();
    }

    /**
     * Inserts a value into the heap.
     */
    public void insert(int value) {
        heap.add(value);
        siftUp(heap.size() - 1);
    }

    /**
     * Removes and returns the maximum element.
     */
    public int popMax() {
        if (heap.isEmpty()) {
            throw new NoSuchElementException("Heap is empty");
        }

        int maxValue = heap.get(0);
        int lastIndex = heap.size() - 1;

        if (lastIndex == 0) {
            heap.remove(lastIndex);
            return maxValue;
        }

        heap.set(0, heap.get(lastIndex));
        heap.remove(lastIndex);

        siftDown(0);

        return maxValue;
    }

    /**
     * Returns the maximum element without removing it.
     */
    public int peekMax() {
        if (heap.isEmpty()) {
            throw new NoSuchElementException("Heap is empty");
        }

        return heap.get(0);
    }

    /**
     * Returns heap size.
     */
    public int size() {
        return heap.size();
    }

    /**
     * Returns true if heap is empty.
     */
    public boolean isEmpty() {
        return heap.isEmpty();
    }

    /**
     * Moves element upward until heap property is restored.
     */
    private void siftUp(int index) {
        while (index > 0) {
            int parentIndex = (index - 1) / branchingFactor;

            if (heap.get(index) <= heap.get(parentIndex)) {
                break;
            }

            swap(index, parentIndex);
            index = parentIndex;
        }
    }

    /**
     * Moves element downward until heap property is restored.
     */
    private void siftDown(int index) {
        int size = heap.size();

        while (true) {
            int largestIndex = index;

            int firstChild = index * branchingFactor + 1;

            for (int offset = 0; offset < branchingFactor; offset++) {
                int childIndex = firstChild + offset;

                if (childIndex >= size) {
                    break;
                }

                if (heap.get(childIndex) > heap.get(largestIndex)) {
                    largestIndex = childIndex;
                }
            }

            if (largestIndex == index) {
                break;
            }

            swap(index, largestIndex);
            index = largestIndex;
        }
    }

    /**
     * Swaps two heap elements.
     */
    private void swap(int firstIndex, int secondIndex) {
        int temp = heap.get(firstIndex);
        heap.set(firstIndex, heap.get(secondIndex));
        heap.set(secondIndex, temp);
    }

    /**
     * Simple test.
     */
    public static void main(String[] args) {

        // x = 2 => each parent has 2^2 = 4 children
        PowerOfTwoMaxHeap heap = new PowerOfTwoMaxHeap(2);

        heap.insert(10);
        heap.insert(40);
        heap.insert(15);
        heap.insert(70);
        heap.insert(20);
        heap.insert(100);
        heap.insert(5);

        while (!heap.isEmpty()) {
            System.out.print(heap.popMax() + " ");
        }

        // Output:
        // 100 70 40 20 15 10 5
    }
}